File: frontend/src/types/mcp.ts
import type { Scope, OverrideInfo } from './common';

export interface McpServerConfig {
  command: string;
  args?: string[];
  env?: Record<string, string>;
  url?: string;
  transport?: string;  // "stdio", "http", "sse"
}

export interface McpServer {
  name: string;
  config: McpServerConfig;
  scope: Scope;
  active: boolean;
  overridden?: OverrideInfo;
  source?: string;
}

export interface McpServerResponse {
  name: string;
  config: McpServerConfig;
  scope: Scope;
  active: boolean;
  overridden?: Record<string, unknown> | null;
  source?: string;
}

export interface CreateMcpRequest {
  name: string;
  command?: string;
  args?: string[];
  env?: Record<string, string>;
  scope: Scope;
  url?: string;
  transport?: string;
}

export interface UpdateMcpRequest {
  command?: string;
  args?: string[];
  env?: Record<string, string>;
  scope: Scope;
  url?: string;
  transport?: string;
}

export type ConnectivityStatus = 'unknown' | 'testing' | 'connected' | 'error';

export interface ConnectivityResult {
  name: string;
  status: ConnectivityStatus;
  message?: string;
  latency?: number;
}

export interface McpConflict {
  name: string;
  project_config: McpServerConfig;
  global_config: McpServerConfig;
}

File: backend/models/mcp.py
from pydantic import BaseModel
from typing import Optional, Dict, List

from .common import Scope


class McpServerConfig(BaseModel):
    command: str = ""
    args: Optional[List[str]] = []
    env: Optional[Dict[str, str]] = {}
    url: Optional[str] = None
    transport: Optional[str] = None  # "stdio", "http", "sse"


class CreateMcpRequest(BaseModel):
    name: str
    command: str = ""
    args: Optional[List[str]] = []
    env: Optional[Dict[str, str]] = {}
    scope: Scope = "project"
    url: Optional[str] = None
    transport: Optional[str] = None


class UpdateMcpRequest(BaseModel):
    command: str = ""
    args: Optional[List[str]] = []
    env: Optional[Dict[str, str]] = {}
    scope: Scope = "project"
    url: Optional[str] = None
    transport: Optional[str] = None


class McpServerResponse(BaseModel):
    name: str
    config: McpServerConfig
    scope: Scope
    active: bool = True
    overridden: Optional[dict] = None
    source: Optional[str] = None


class ConnectivityResult(BaseModel):
    name: str
    status: str  # unknown, testing, connected, error
    message: Optional[str] = None
    latency: Optional[float] = None


class McpConflict(BaseModel):
    name: str
    project_config: McpServerConfig
    global_config: McpServerConfig
